storage/src/snapshot_cache.rs

```diff
@@ -0,0 +1,268 @@
+use crate::Snapshot;
+use std::collections::VecDeque;
+
+use anyhow::anyhow;
+
+/// The number of snapshots kept in the [`SnapshotCache`].
+pub const DEFAULT_SNAPSHOT_CACHE_SIZE: usize = 10;
+
+/// A circular cache for storing [`Snapshot`]s.
+///
+/// # Usage
+///
+/// [`Snapshot`]s are inserted in the cache using the [`push`] or [`try_push`]
+/// methods. If the cache is full, the oldest entry will be evicted to make space
+/// for the newer entry.
+///
+/// # Constraints
+///
+/// [`Snapshot`]s must be inserted sequentially relative to their [`jmt::Version`]
+/// numbers, and have consecutive version numbers.
+///
+/// # Special case
+///
+/// The first `Snapshot` inserted is allowed to have a "pre-genesis" version number
+/// of [`u64::MAX`], but every subsequent inserts must have consecutive versions.
+pub struct SnapshotCache {
+    /// A sequence of increasingly recent [`Snapshot`]s.
+    cache: VecDeque<Snapshot>,
+    /// The max length and capacity of the cache.
+    max_size: usize,
+}
+
+impl SnapshotCache {
+    /// Creates a [`SnapshotCache`] with `max_size` capacity, and inserts an initial `Snapshot` in
+    /// it.
+    ///
+    /// ## Panics
+    ///
+    /// This function panics if the specified capacity is zero.
+    pub fn new(initial: Snapshot, max_size: usize) -> Self {
+        if max_size == 0 {
+            panic!("snapshot_cache: capacity must be greater than 0.");
+        }
```

Comment:
Why panic here? Do we really want to kill the program over that? Why not do
```suggestion
    pub fn new(initial: Snapshot, max_size: usize) -> Self {
        let max_size = std::cmp::max(max_size, 1);
```
instead?